kerastuner/tuners/ultraband.py

```diff
@@ -37,3 +229,55 @@ def __init__(self,
             objective,
             max_trials,
             **kwargs)
+
+    def run_trial(self, trial, hp, fit_args, fit_kwargs):
+        fit_kwargs = copy.copy(fit_kwargs)
+        original_callbacks = fit_kwargs.get('callbacks', [])[:]
+        for i in range(self.executions_per_trial):
+            execution_id = tuner_utils.format_execution_id(
+                i, self.executions_per_trial)
+            # Patch fit arguments
+            max_epochs, max_steps = tuner_utils.get_max_epochs_and_steps(
+                fit_args, fit_kwargs)
+            fit_kwargs['verbose'] = 0
+
+            # Get model; this will reset the Keras session
+            if not self.tune_new_entries:
+                hp = hp.copy()
+            model = self._build_model(hp)
+            self._compile_model(model)
+
+            # Start execution
+            execution = execution_module.Execution(
+                execution_id=execution_id,
+                trial_id=trial.trial_id,
+                max_epochs=max_epochs,
+                max_steps=max_steps,
+                base_directory=trial.directory)
+            trial.executions.append(execution)
+            self.on_execution_begin(trial, execution, model)
+
+            # During model `fit`,
+            # the patched callbacks call
+            # `self.on_epoch_begin`, `self.on_epoch_end`,
+            # `self.on_batch_begin`, `self.on_batch_end`.
+            fit_kwargs['callbacks'] = self._inject_callbacks(
+                original_callbacks, trial, execution)
+            if 'epochs' in hp.values:
+                fit_kwargs['epochs'] = hp.values['epochs']
+            if 'trial_id' in hp.values:
```

Comment:
Here's a simpler way to handle this:

- no `trial_id` in hp
- the tuner computes a hash of `hp.values` in `on_trial_begin`.
- the tuner maintains a map {hash: latest_checkpoint} for these hashes.
- if there's a match in the map, reload the checkpoint (in `on_execution_begin`).

Reply:
For the record, we override `on_execution_begin` to load the weights using `tuner/trial_id` in `trial.hyperparameters`. We also calls the super `on_execution_begin`.